karamel-core/src/main/java/se/kth/karamel/backend/machines/SshMachine.java

```diff
@@ -126,41 +127,108 @@ private boolean anyFailure() {
         try {
           if (machineEntity.getLifeStatus() == MachineRuntime.LifeStatus.CONNECTED
               && (machineEntity.getTasksStatus() == MachineRuntime.TasksStatus.ONGOING
               || machineEntity.getTasksStatus() == MachineRuntime.TasksStatus.EMPTY)) {
             try {
+              boolean retry = false;
               if (activeTask == null) {
                 if (taskQueue.isEmpty()) {
                   machineEntity.setTasksStatus(MachineRuntime.TasksStatus.EMPTY, null, null);
                 }
                 activeTask = taskQueue.take();
                 logger.debug(String.format("%s: Taking a new task from the queue.", machineEntity.getId()));
                 machineEntity.setTasksStatus(MachineRuntime.TasksStatus.ONGOING, null, null);
+                hasAbortedDueToFailure = false;
               } else {
+                retry = true;
                 logger.debug(
                     String.format("%s: Retrying a task that didn't complete on last execution attempt.",
                         machineEntity.getId()));
               }
               logger.debug(String.format("%s: Task for execution.. '%s'", machineEntity.getId(), activeTask.getName()));
+
+              Optional<RecipeSerialization> recipeSerialization = getRecipeSerialization(activeTask);
+              if (recipeSerialization.isPresent()) {
+                RecipeSerialization rs = recipeSerialization.get();
+                // In retries we don't check if it has failed because we already know it has failed, hence retried
+                // If we do check, it will lead to deadlock
+                //
+                // After the task gotten the green light to proceed with running the task - parallelism claim
+                // has been satisfied, we check again if the task has failed. While waiting for the claim to be
+                // satisfied, the task might have failed on another node. In that case, we continue the loop without
+                // executing the task. Then we come here, where it is a retry but it has aborted due to failure
+                // and we wait until the failure is resolved
+                if (!retry || hasAbortedDueToFailure) {
+                  logger.debug(String.format("%s: retry: %s hasAborted: %s", machineEntity.getId(),
+                      activeTask.getName(), retry, hasAbortedDueToFailure));
+                  synchronized (rs) {
+                    logger.debug(String.format("%s: Recipe %s RecipeSerializationID: %s Has failed: %s",
+                        activeTask.getMachine().getId(),
+                        activeTask.getName(),
+                        rs.hashCode(),
+                        rs.hasFailed()));
+                    if (rs.hasFailed()) {
+                      logger.info(String.format("%s: Recipe %s has failed on another node. Wait until it succeeds",
+                          machineEntity.getId(), activeTask.getName()));
+                      rs.wait();
```

Comment:
would we be blocked here if there is only a single recipe and we mistakenly added the parallelism config to `1`?